src/libsyntax/parse/parser.rs

```diff
@@ -4726,9 +4726,13 @@ impl<'a> Parser<'a> {
             let fields = try!(self.parse_record_struct_body(&class_name));
             (fields, None)
         // Tuple-style struct definition with optional where-clause.
-        } else {
+        } else if self.token == token::OpenDelim(token::Paren) {
             let fields = try!(self.parse_tuple_struct_body(&class_name, &mut generics));
             (fields, Some(ast::DUMMY_NODE_ID))
+        } else {
+            let token_str = self.this_token_to_string();
+            return Err(self.fatal(&format!("expected `where`, `{}`, `(`, or `;` after struct \
+                                            name, found `{}`", "{", token_str)))
```

Comment:
I don't really know for sure, I just moved this piece of code from `parse_tuple_struct_body`, here: https://github.com/matklad/rust/blob/3a360fca78b415d31783ec715a2c9024a89eeb75/src/libsyntax/parse/parser.rs#L4806-L4807 :)

If I just remove this explicit check and if test, then everything works, but the error message for one tests changes and the build fails.